apps/emqx/src/emqx_limiter/src/emqx_limiter.erl

```diff
@@ -211,20 +171,34 @@ delete_group(Group) ->
 %% Zone config update
 %%--------------------------------------------------------------------
 
-add_handler() ->
-    ok = emqx_config_handler:add_handler([mqtt, limiter], ?MODULE),
-    ok.
+-spec post_zone_config_update(emqx_config:config(), emqx_config:config()) -> ok.
+post_zone_config_update(OldZoneConfig, NewZoneConfig) ->
+    do_post_zone_config_update(is_initialized(), OldZoneConfig, NewZoneConfig).
 
-remove_handler() ->
-    ok = emqx_config_handler:remove_handler([mqtt, limiter]),
+do_post_zone_config_update(true, OldZoneConfig, NewZoneConfig) ->
+    #{
+        added := Added,
+        removed := Removed,
+        changed := Changed
+    } = emqx_utils_maps:diff_maps(NewZoneConfig, OldZoneConfig),
+
+    ok = maps:foreach(
+        fun(Zone, ZoneConfig) ->
+            create_limiters_for_zone(Zone, limiter_config_for_zone(ZoneConfig))
+        end,
+        Added
+    ),
+    ok = maps:foreach(
+        fun(Zone, {_OldZoneConfig, ZoneConfig}) ->
+            update_limiters_for_zone(Zone, limiter_config_for_zone(ZoneConfig))
+        end,
+        Changed
+    ),
+    %% Never delete the default zone
```

Comment:
Q: if the `default` zone _is indeed_ deleted (for whatever reason), shouldn't we also delete its limiters to avoid inconsistencies or leaking resources?

Reply:
We do not have any logic that forbids the removal of used zones, so I left it to avoid the most common probable errors.
We shouldn't hit resource leakage, we preserve a few small values of constant size.